x-pack/solutions/search/packages/kbn-search-queries/src/fields.ts

```diff
@@ -0,0 +1,178 @@
+  const indexFields = Object.keys(fieldDescriptors).reduce<IndexFields>(
+    (acc: IndexFields, index: string) => {
+      const indexFieldDescriptors = fieldDescriptors[index];
+
+      // if there are no source fields, we don't need to suggest anything
+      if (indexFieldDescriptors.source_fields.length === 0) {
+        return {
+          ...acc,
+          [index]: [],
+        };
+      }
+
+      return {
+        ...acc,
+        [index]: indexFieldDescriptors.source_fields,
+      };
+    },
+    {}
+  );
+
+  return indexFields;
+}
+
+export function getIndicesWithNoSourceFields(fields: IndicesQuerySourceFields): string | undefined {
+  const defaultSourceFields = getDefaultSourceFields(fields);
+  const indices = Object.keys(defaultSourceFields).reduce<string[]>((result, index: string) => {
+    if (defaultSourceFields[index].length === 0) {
+      result.push(index);
+    }
+
+    return result;
+  }, []);
+
+  return indices.length === 0 ? undefined : indices.join();
+}
+
+export function getDefaultQueryFields(fieldDescriptors: IndicesQuerySourceFields): IndexFields {
+  const indexFields = Object.keys(fieldDescriptors).reduce<IndexFields>(
+    (acc: IndexFields, index: string) => {
+      const indexFieldDescriptors = fieldDescriptors[index];
+      const fields: string[] = [];
+
+      if (indexFieldDescriptors.semantic_fields.length > 0) {
+        fields.push(...indexFieldDescriptors.semantic_fields.map((x) => x.field));
+      } else if (indexFieldDescriptors.elser_query_fields.length > 0) {
+        const suggested = indexFieldDescriptors.elser_query_fields.filter((x) =>
+          SUGGESTED_SPARSE_FIELDS.includes(x.field)
+        );
+        if (suggested.length > 0) {
+          fields.push(...suggested.map((x) => x.field));
+        } else {
+          fields.push(...indexFieldDescriptors.elser_query_fields.map((x) => x.field));
+        }
+      } else if (indexFieldDescriptors.dense_vector_query_fields.length > 0) {
+        const suggested = indexFieldDescriptors.dense_vector_query_fields.filter((x) =>
+          SUGGESTED_DENSE_VECTOR_FIELDS.includes(x.field)
+        );
+
+        if (suggested.length > 0) {
+          fields.push(...suggested.map((x) => x.field));
+        } else {
+          fields.push(...indexFieldDescriptors.dense_vector_query_fields.map((x) => x.field));
+        }
+      } else if (indexFieldDescriptors.bm25_query_fields.length > 0) {
+        const suggested = indexFieldDescriptors.bm25_query_fields.filter((x) =>
+          SUGGESTED_BM25_FIELDS.includes(x)
+        );
+        if (suggested.length > 0) {
+          fields.push(...suggested);
+        } else {
+          fields.push(indexFieldDescriptors.bm25_query_fields[0]);
+        }
+      }
+
+      return {
+        ...acc,
+        [index]: fields,
+      };
+    },
+    {}
+  );
+
+  return indexFields;
+}
+
+export function buildFieldDescriptorForIndex(
```

Comment:
this needs some more work as it is pretty minimal and only support top-level properties.